audio/out/ao_oss.c

```diff
@@ -0,0 +1,422 @@
+            if (!ao_chmap_sel_get_def(ao, &sel, &channels, nchannels))
+                goto fail;
+        }
+        MP_VERBOSE(ao, "Using %d channels (requested: %d).\n",
+                   channels.num, reqchannels);
+    }
+
+    /* Sample rate. */
+    samplerate = ao->samplerate;
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_SPEED, &samplerate) == -1)
+        goto err_out_ioctl;
+    MP_VERBOSE(ao, "Using %d Hz samplerate.\n", samplerate);
+
+    /* Get buffer size. */
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_GETOSPACE, &info) == -1)
+        goto err_out_ioctl;
+    /* See ao.c ao->sstride initializations and get_state(). */
+    ao->device_buffer = ((info.fragstotal * info.fragsize) /
+        af_fmt_to_bytes(format));
+    if (!af_fmt_is_planar(format)) {
+        ao->device_buffer /= channels.num;
+    }
+
+    /* Update sound params. */
+    ao->format = format;
+    ao->samplerate = samplerate;
+    ao->channels = channels;
+    p->bps = (channels.num * samplerate * af_fmt_to_bytes(format));
+
+    reset(ao);
+
+    return 0;
+
+err_out_ioctl:
+    MP_WARN_IOCTL_ERR(ao);
+fail:
+    uninit(ao);
+    return -1;
+}
+
+static void uninit(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    if (p->dsp_fd == -1)
+        return;
+    ioctl(p->dsp_fd, SNDCTL_DSP_RESET, NULL);
+    close(p->dsp_fd);
+    p->dsp_fd = -1;
+    p->playing = false;
+}
+
+static int control(struct ao *ao, enum aocontrol cmd, void *arg)
+{
+    struct priv *p = ao->priv;
+    ao_control_vol_t *vol = (ao_control_vol_t *)arg;
+    int v;
+
+    if (p->dsp_fd < 0)
+        return CONTROL_ERROR;
+
+    switch (cmd) {
+    case AOCONTROL_GET_VOLUME:
+        if (ioctl(p->dsp_fd, SNDCTL_DSP_GETPLAYVOL, &v) == -1) {
+            MP_WARN_IOCTL_ERR(ao);
+            return CONTROL_ERROR;
+        }
+        vol->right = ((v & 0xff00) >> 8);
+        vol->left = (v & 0x00ff);
+        return CONTROL_OK;
+    case AOCONTROL_SET_VOLUME:
+        v = ((int)vol->right << 8) | (int)vol->left;
+        if (ioctl(p->dsp_fd, SNDCTL_DSP_SETPLAYVOL, &v) == -1) {
+            MP_WARN_IOCTL_ERR(ao);
+            return CONTROL_ERROR;
+        }
+        return CONTROL_OK;
+    }
+
+    return CONTROL_UNKNOWN;
+}
+
+static void reset(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    p->playing = false;
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_RESET, NULL) == 0) {
+        set_pause(ao, true);
+    } else {
+        MP_WARN_IOCTL_ERR(ao);
+        if (p->resetting)
+            return; /* Avoid infinite loop: init() -> reset() -> init()... */
+        p->resetting = true;
+        MP_WARN(ao, "Force reinitialize audio device.\n");
+        uninit(ao);
+        init(ao);
+        p->resetting = false;
+    }
+}
+
+static bool set_pause(struct ao *ao, bool pause)
+{
+    struct priv *p = ao->priv;
+    int trig = 0; /* Both directions disabled */
+
+    if (!pause) {
+        trig |= (PCM_ENABLE_INPUT | PCM_ENABLE_OUTPUT); /* Both directions enabled */
```

Comment:
```PCM_ENABLE_INPUT``` - copy-paste from man, will remove it.

Why ```SNDCTL_DSP_HALT``` should be used instead of ```SNDCTL_DSP_SETTRIGGER```? Because write required?
```SNDCTL_DSP_RESET``` = ```SNDCTL_DSP_HALT``` already used in reset to clear buffer.

Reply:
SNDCTL_DSP_SETTRIGGER was never intended to be used to stop a stream after the playback has been started. It is mainly used to start the stream synchronously with the recording direction. Actually, if SNDCTL_DSP_SETTRIGGER isn't used at all, as long as you write something into the streaming direction for the first time, the playback would start.

Reply:
A quote from http://manuals.opensound.com/developer/SNDCTL_DSP_SETTRIGGER.html:
> This ioctl call cannot be used to stop and restart the device after the playback has been started (this may work with some OSS implementations but fail wit the the rest of them). It is necessary to use the SNDCTL_DSP_HALT family for this purpose.